File: heap/30_maximum_product_after_k_increments.cpp
#include <bits/stdc++.h>
using namespace std;

/*
Problem Description:
Given nums and k operations, in each operation increment one element by 1.
Return maximum possible product of array after k operations.

Sample Input/Output:
1) nums=[0,4], k=5 -> 20
2) nums=[6,3,3,2], k=2 -> 216
3) nums=[1,1,1], k=3 -> 8

Code Explanation:
To maximize product, always increment smallest element.
Use min-heap for k operations, then multiply all values.
*/

long long maximumProduct(vector<int> &nums, int k) {
    const long long MOD = 1'000'000'007;
    priority_queue<long long, vector<long long>, greater<long long>> minHeap;
    for (int x : nums) minHeap.push(x);

    while (k--) {
        long long x = minHeap.top();
        minHeap.pop();
        minHeap.push(x + 1);
    }

    long long product = 1;
    while (!minHeap.empty()) {
        product = (product * minHeap.top()) % MOD;
        minHeap.pop();
    }
    return product;
}

int main() {
    vector<int> a1 = {0,4};
    vector<int> a2 = {6,3,3,2};
    vector<int> a3 = {1,1,1};

    cout << "Output 1: " << maximumProduct(a1, 5) << "\n";
    cout << "Output 2: " << maximumProduct(a2, 2) << "\n";
    cout << "Output 3: " << maximumProduct(a3, 3) << "\n";
    return 0;
}

File: heap/30_maximum_product_after_k_increments.py
"""
PROBLEM: Maximum Product After K Increments (LeetCode 2233)
DIFFICULTY: Medium
PATTERN: Heap - Min-Heap greedy (always increment smallest)

DESCRIPTION:
Given nums and k operations, in each operation increment one element by 1.
Return maximum possible product modulo 10^9 + 7.

EXAMPLE:
Input: nums=[0,4], k=5 -> 20
Input: nums=[6,3,3,2], k=2 -> 216
Input: nums=[1,1,1], k=3 -> 8

TIME COMPLEXITY: O((n + k) log n)
SPACE COMPLEXITY: O(n)
"""

import heapq
from typing import List


class Solution:
    def maximumProduct(self, nums: List[int], k: int) -> int:
        MOD = 10 ** 9 + 7
        heapq.heapify(nums)

        for _ in range(k):
            smallest = heapq.heappop(nums)
            heapq.heappush(nums, smallest + 1)

        result = 1
        for x in nums:
            result = (result * x) % MOD
        return result


if __name__ == "__main__":
    sol = Solution()
    test_cases = [
        ([0, 4], 5),
        ([6, 3, 3, 2], 2),
        ([1, 1, 1], 3),
    ]
    for nums, k in test_cases:
        print(f"nums={nums}, k={k} -> {sol.maximumProduct(nums[:], k)}")
